src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/aws/AWSEC2JCloudsProvider.java

```diff
@@ -303,6 +312,28 @@ public void removeInstancePublicIp(Infrastructure infrastructure, String instanc
         }
     }
 
+    @Override
+    public void deleteKeyPair(Infrastructure infrastructure, String keyPairName, String region) {
+        KeyPairApi keyPairApi = getKeyPairApi(infrastructure);
+        keyPairApi.deleteKeyPairInRegion(region, keyPairName);
+        log.info("Removed the key pair [{}] in the region [{}]", keyPairName, region);
+    }
+
+    @Override
+    public void deleteInfrastructure(Infrastructure infrastructure) {
+        // if the infrastructure has used the auto-generated security group, remove it.
+        if (usingAutoGeneratedSecurityGroupInfrastructures.contains(infrastructure.getId())) {
+            String securityGroupName = getAutoGeneratedSecurityGroupName(infrastructure.getId());
+            SecurityGroupApi securityGroupApi = getSecurityGroupApi(infrastructure);
+            securityGroupApi.deleteSecurityGroupInRegion(infrastructure.getRegion(), securityGroupName);
+            log.info(String.format("Removed the auto-generated security group [%s] for the infrastructure [%s]",
+                                   securityGroupName,
+                                   infrastructure.getId()));
+        }
+        super.deleteInfrastructure(infrastructure);
```

Comment:
Shouldn't this line be invoked before the removal of SG & keyPair ?

Reply:
As the removal of SG & keyPair needs jclouds infrastructure computerService (to get the corresponding API client), which is the removed during `super.deleteInfrastructure`, so it must be done at the end.